### PR TITLE
drivers: Add Silicon Labs Si1133 sensor driver.

### LINKED ISSUE #35354: Adding support for measurement of Ultraviolet(UV) Light.
The Zephyr sensor subsystem currently does not support the measurement of Ultraviolet(UV) light.

This can be solved by adding a new enumeration entry in the header file "include/drivers/sensor.h" so that sensors that measure UV light(such as Silicon Labs Si1133) can report UV measurements on this UV channel.

The following two lines are to be added to the  "enum sensor_channel", which is present in the above header file.
/** Ultraviolet index, in UVI. */
SENSOR_CHAN_UV,

### PR DESCRIPTION
**Pull request by ghost**
This driver supports the Silicon Labs Si1133 UV Index and Ambient Light Sensor.

Tested on Silicon Labs Thunderboard Sense 2(SLTB004A) board.

Origin: Original

Fixes #35354

Signed-off-by: Murali Tejeshwar Janaswami <muralitejeshwar.janaswami@gmail.com>

### CODE REVIEW

File: include/drivers/sensor.h [hunk 1]
```diff
@@ -90,6 +90,8 @@ enum sensor_channel {
 	SENSOR_CHAN_HUMIDITY,
 	/** Illuminance in visible spectrum, in lux. */
 	SENSOR_CHAN_LIGHT,
+	/** Ultraviolet index, in UVI. */
+	SENSOR_CHAN_UV,
```
Comment (JordanYates) [inline include/drivers/sensor.h:94]: Same comment from the previous PR:
> I wouldn't put this in the middle of the other light channels given that the units are not the same.
Its also not obvious that SENSOR_CHAN_UV and SENSOR_CHAN_IV are returning very different things.
Perhaps SENSOR_CHAN_UV_INDEX?

File: drivers/sensor/si1133/Kconfig [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+config SI1133
+	bool "Si1133 UV Index and Ambient Light Sensor"
+	depends on I2C
+	help
+	  Enable driver for Si1133 UV Index and Ambient Light Sensor.
+
+if SI1133
+
+choice SI1133_OPERATING_MODE
```
Comment (JordanYates) [inline drivers/sensor/si1133/Kconfig:14]: Why is this choice needed when the reference driver from mbed works fine (as far as I can tell) across all light levels?

File: drivers/sensor/si1133/si1133.c [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+}
+
+static int si1133_channel_get(const struct device *dev,
+			      enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct si1133_data *drv_data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_LIGHT:
```
Comment (JordanYates) [inline drivers/sensor/si1133/si1133.c:282]: `SENSOR_CHAN_LIGHT` and `SENSOR_CHAN_IR` are documented to return values as Lux, whereas it appears you are returning the raw photodiode ADC readings. These are not the same thing.

File: drivers/sensor/si1133/si1133.c [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+	    (data[3] << 16) | (data[4] << 8) | data[5];
+	uv_light_raw = (data[6] << 8) | data[7];
+#else
+	drv_data->data.visible_light = (data[0] << 8) | data[1];
+	drv_data->data.infrared_light = (data[2] << 8) | data[3];
+	uv_light_raw = (data[4] << 8) | data[5];
+#endif
+	/* convert the sensor's raw UV measurement into UVI scale */
+	drv_data->data.ultraviolet_light =
+	    0.0187 * ((0.00391 * SQUARE(uv_light_raw)) + uv_light_raw);
```
Comment (JordanYates) [inline drivers/sensor/si1133/si1133.c:262]: Where do these magic numbers come from?

**Review by JordanYates — CHANGES_REQUESTED** [review]
The Visible light and IR channels need to return the correct value types.
This implementation is quite different from what I would consider to be the reference driver from Silicon Labs, particularly around the channel setup and data conversion: https://os.mbed.com/teams/SiliconLabs/code/Si1133//file/f780ca9105bb/Si1133.cpp/.

File: include/drivers/sensor.h (hunk 1 above)
Comment (murali-tejeshwar) [reply include/drivers/sensor.h:94]: UV is always measured in the UVI scale(see https://en.wikipedia.org/wiki/Ultraviolet_index).

Yes, the placement of the channel is not quite right.

File: drivers/sensor/si1133/Kconfig (hunk 2 above)
Comment (murali-tejeshwar) [reply drivers/sensor/si1133/Kconfig:14]: Sunlight has much higher lux values than indoor lighting. 16-bit unsigned variables will not be sufficient for holding values larger than 65535. This operating mode determines if the sensor is to be operated in 24-bit mode(which can store larger values) or 16-bit mode(which is sufficient for low light scenarios).

File: drivers/sensor/si1133/si1133.c (hunk 4 above)
Comment (murali-tejeshwar) [reply drivers/sensor/si1133/si1133.c:262]: Page 9 from this document: https://www.silabs.com/documents/public/application-notes/AN968-Si1133-UV-Index-Sensor-Electrical-and-Optical-Design-Guide.pdf

File: drivers/sensor/si1133/Kconfig (hunk 2 above)
Comment (JordanYates) [reply drivers/sensor/si1133/Kconfig:14]: Right, but the reference driver configures 2 photodiodes and chooses which one to use for the output based on the saturation of the more sensitive photodiode. Why can't we do the same thing here?

File: include/drivers/sensor.h (hunk 1 above)
Comment (JordanYates) [reply include/drivers/sensor.h:94]: I'm not sure that's strictly accurate, sure that's the normal day-to-day usage, but I would think irradiance would be the normal scientific term. The current definition for `SENSOR_CHAN_IR` is obviously wrong though, as lux only has meaning in visible wavelengths. A more consistent solution would be:
```
SENSOR_CHAN_IR, /* Irradiance, W/m^2 */
SENSOR_CHAN_UV, /* Irradiance, W/m^2 */
SENSOR_CHAN_UV_INDEX /* Ultraviolet Index, unitless */
```

File: drivers/sensor/CMakeLists.txt [hunk 5]
```diff
@@ -75,6 +75,7 @@ add_subdirectory_ifdef(CONFIG_SGP40		sgp40)
 add_subdirectory_ifdef(CONFIG_SHTCX		shtcx)
 add_subdirectory_ifdef(CONFIG_SHT3XD		sht3xd)
 add_subdirectory_ifdef(CONFIG_SHT4X		sht4x)
+add_subdirectory_ifdef(CONFIG_SI1133		si1133)
```
Comment (chrta) [inline drivers/sensor/CMakeLists.txt:78]: The sensor should probably be added to `tests/drivers/build_all/sensor` so the driver is built on the buildserver.

**Review by chrta — CHANGES_REQUESTED** [review]
Thank you for the contribution. This is just my initial brief review. In addition it would be good if there would be some sample checked in that runs on the Silicon Labs Thunderboard Sense 2 board that activates and uses the sensor. This way i could test this on my board and the buildserver could always check that this example builds.

**Pull request closed without merging**